### PR TITLE
feat: implementation of basic access control for Files API

### PR DESCRIPTION
**Pull request by Maxim-Gadalov**

### CODE REVIEW

File: src/main/java/com/epam/aidial/core/controller/DownloadFileController.java [hunk 1]
```diff
@@ -36,18 +31,20 @@ public class DownloadFileController {
...
      */
-    public Future<?> download(String path) {
-        String pathType = context.getRequest().params().get(PATH_TYPE_QUERY_PARAMETER);
-        String absoluteFilePath;
-        if (RELATIVE_PATH_TYPE.equals(pathType)) {
-            absoluteFilePath = BlobStorageUtil.buildAbsoluteFilePath(context, path);
-        } else {
-            absoluteFilePath = path;
+    public Future<?> download(String bucket, String filePath) {
+        String expectedUserBucket = BlobStorageUtil.buildUserBucket(context);
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/controller/DownloadFileController.java]: The same piece of code is duplicated across FILE API. I can suggest introducing AbstractFilleController to move this logic there and call abstract method which will be overriden in subclasses.

File: src/main/java/com/epam/aidial/core/data/FileMetadataBase.java [hunk 2]
```diff
@@ -9,6 +9,8 @@
 @NoArgsConstructor
 public abstract class FileMetadataBase {
     private String name;
-    private String path;
+    private String parentPath;
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/data/FileMetadataBase.java:12]: Can we rename to folder?

File: src/main/java/com/epam/aidial/core/security/EncryptionService.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    private final SecretKey key;
+    private final IvParameterSpec iv = new IvParameterSpec(
+            new byte[]{25, -13, -25, -119, -42, 117, -118, -128, -101, 20, -103, -81, -48, -23, -54, -113});
+
+    public EncryptionService(Encryption config) {
+        this(config.getPassword(), config.getSalt());
+    }
+
+    EncryptionService(String password, String salt) {
+        try {
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/security/EncryptionService.java:33]: Can you move checks in the beginning?

File: src/main/java/com/epam/aidial/core/security/EncryptionService.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    public String decrypt(String value) {
+        try {
+            Cipher cipher = Cipher.getInstance(TRANSFORMATION);
+            cipher.init(Cipher.DECRYPT_MODE, key, iv);
+            return new String(cipher.doFinal(Base64.getDecoder()
+                    .decode(value)));
+        } catch (Exception e) {
+            log.error("Failed to decrypt value " + value, e);
+
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/security/EncryptionService.java]: remove empty line

File: src/main/java/com/epam/aidial/core/security/EncryptionService.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            Cipher cipher = Cipher.getInstance(TRANSFORMATION);
+            cipher.init(Cipher.ENCRYPT_MODE, key, iv);
+            return Base64.getEncoder()
+                    .encodeToString(cipher.doFinal(value.getBytes()));
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public String decrypt(String value) {
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/security/EncryptionService.java:53]: add @Nullable

File: src/main/java/com/epam/aidial/core/storage/ResourceType.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package com.epam.aidial.core.storage;
+
+public enum ResourceType {
+    FILES("files");
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/storage/ResourceType.java]: nit: FILE? Why plural?

**Review by artsiomkorzun — CHANGES_REQUESTED** [review]

**Comment by avavilau** [comment]
/deploy-review
>Deployment status: success

**Comment by vladisavvv** [comment]
/deploy-review
>Deployment status: success

File: src/main/java/com/epam/aidial/core/controller/DownloadFileController.java (hunk 1 above)
Comment (Maxim-Gadalov) [reply src/main/java/com/epam/aidial/core/controller/DownloadFileController.java]: fixed

File: src/main/java/com/epam/aidial/core/data/FileMetadataBase.java (hunk 2 above)
Comment (Maxim-Gadalov) [reply src/main/java/com/epam/aidial/core/data/FileMetadataBase.java:12]: it's not a folder, it's a path, like folder1/folder2/folder3

File: src/main/java/com/epam/aidial/core/security/EncryptionService.java (hunk 3 above)
Comment (Maxim-Gadalov) [reply src/main/java/com/epam/aidial/core/security/EncryptionService.java:33]: fixed

File: src/main/java/com/epam/aidial/core/security/EncryptionService.java (hunk 4 above)
Comment (Maxim-Gadalov) [reply src/main/java/com/epam/aidial/core/security/EncryptionService.java]: fixed

File: src/main/java/com/epam/aidial/core/security/EncryptionService.java (hunk 5 above)
Comment (Maxim-Gadalov) [reply src/main/java/com/epam/aidial/core/security/EncryptionService.java:53]: added

File: src/main/java/com/epam/aidial/core/storage/ResourceType.java (hunk 6 above)
Comment (Maxim-Gadalov) [reply src/main/java/com/epam/aidial/core/storage/ResourceType.java]: fixed

File: src/main/java/com/epam/aidial/core/controller/DeleteFileController.java [hunk 7]
```diff
@@ -3,26 +3,24 @@
...
-    private final ProxyContext context;
+public class DeleteFileController extends AccessControlBaseController {
+
+    public DeleteFileController(Proxy proxy, ProxyContext context) {
+        super(proxy, context);
+    }
+
+    @Override
+    protected Future<?> handle(String bucketName, String bucketLocation, String filePath) {
+        ResourceDescription resource = ResourceDescription.from(ResourceType.FILE, bucketName, bucketLocation, filePath);
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/controller/DeleteFileController.java]: Check that it is a file not a folder.

File: src/main/java/com/epam/aidial/core/controller/DownloadFileController.java [hunk 8]
```diff
@@ -18,36 +18,18 @@
...
-    static final String QUERY_METADATA_QUERY_PARAMETER_VALUE = "metadata";
+    public DownloadFileController(Proxy proxy, ProxyContext context) {
+        super(proxy, context);
+    }
 
-    private final Proxy proxy;
-    private final ProxyContext context;
+    @Override
+    protected Future<?> handle(String bucketName, String bucketLocation, String filePath) {
+        ResourceDescription resource = ResourceDescription.from(ResourceType.FILE, bucketName, bucketLocation, filePath);
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/controller/DownloadFileController.java]: same

File: src/main/java/com/epam/aidial/core/controller/ControllerSelector.java [hunk 9]
```diff
@@ -211,8 +212,8 @@ private static Controller selectPost(Proxy proxy, ProxyContext context, String p
         return null;
     }
 
-    private static Controller selectDelete(Proxy proxy, ProxyContext context, String path) {
-        Matcher match = match(PATTERN_FILES, path);
+    private static Controller selectDelete(Proxy proxy, ProxyContext context, String providedPath, String decodedPath) {
+        Matcher match = match(PATTERN_FILES, providedPath);
```
Comment (artsiomkorzun) [inline src/main/java/com/epam/aidial/core/controller/ControllerSelector.java]: Please write some tests. You use Base64 encoding for buckets. It contains "/". How do you distinguish a slash in a bucket's name and in a path?

Comment (artsiomkorzun) [reply src/main/java/com/epam/aidial/core/controller/ControllerSelector.java]: I would use Base58.

**Comment by mikitabut** [comment]
/deploy-review
>Deployment status: success

File: src/main/java/com/epam/aidial/core/controller/ControllerSelector.java (hunk 9 above)
Comment (Maxim-Gadalov) [reply src/main/java/com/epam/aidial/core/controller/ControllerSelector.java]: changed

File: src/main/java/com/epam/aidial/core/controller/DownloadFileController.java (hunk 8 above)
Comment (Maxim-Gadalov) [reply src/main/java/com/epam/aidial/core/controller/DownloadFileController.java]: fixed

File: src/main/java/com/epam/aidial/core/controller/DeleteFileController.java (hunk 7 above)
Comment (Maxim-Gadalov) [reply src/main/java/com/epam/aidial/core/controller/DeleteFileController.java]: fixed

**Review by artsiomkorzun — APPROVED** [review]

**Pull request merged**